draft-how-does-mutant-work.markdown

```diff
@@ -0,0 +1,431 @@
+
+Now to `run`:
+
+```Ruby
+module Mutant
+  class CLI
+    def self.run(arguments)
+      Runner.call(Env::Bootstrap.call(call(arguments))).success?
+    rescue Error => exception
+      $stderr.puts(exception.message)
+      false
+    end
+  end
+end
+```
+
+Narrowing it down further this is the line we're interested in:
+
+```Ruby
+Runner.call(Env::Bootstrap.call(call(arguments))).success?
+```
+
+That line does pack quite a punch, doesn't it?
+
+Let me reformat this line to make clearer what's happening here:
+
+```Ruby
+Runner.call(
+  Env::Bootstrap.call(
+    call(arguments)
+  )
+).success?
+```
+
+Starting from the inside
+
+```Ruby
+  call(arguments)
+```
+
+translates to:
+
+```Ruby
+CLI.new(arguments).config
+```
+
+This is the part that `Procto` provides.
+
+So:
+
+* we're getting back a config based on our CLI arguments
+* then pass that to `Env::Bootstrap.call`
+* pass whatever comes back from that to our Runner
+
+**This leads us to 2 questions:**
+
+1.) What is the `config` and what is `Env::Bootstrap`?
+
+2.) What does the `Runner` do with this environment.
+
+### Configuration
+
+If you check out the constructor of `Mutant::CLI` you'll see this:
+
+```Ruby
+def initialize(arguments)
+  @config = Config::DEFAULT
+
+  parse(arguments)
+end
+```
+
+Ok, nice and easy. First we'll set the configuration to our defaults, then we're parsing the cli arguments and overwrite our configuration based on this.
+
+What does the default configuration look like?
+For that you'll have to check out the top level library file, so lib/mutant.rb:
+
+
+```Ruby
+module Mutant
+  # Reopen class to initialize constant to avoid dep circle
+  class Config
+    DEFAULT = new(
+      debug:             false,
+      expected_coverage: Rational(1),
+      # snip
+      fail_fast:         false,
+      includes:          EMPTY_ARRAY,
+      # snip
+      kernel:            Kernel,
+      load_path:         $LOAD_PATH,
+      matcher:           Matcher::Config::DEFAULT,
+      open3:             Open3,
+      pathname:          Pathname,
+      requires:          EMPTY_ARRAY,
+      reporter:          Reporter::CLI.build($stdout),
+      zombie:            false
+    )
+  end # Config
+end # Mutant
+```
+
+I did remove a lot of the config object above to keep it readable. I just want to highlight a couple of lines I find interesting:
+
+```Ruby
+expected_coverage: Rational(1)
+```
+
+This means that the expected coverage is 100%. Yep, 100%.
+
+```Ruby
+reporter:          Reporter::CLI.build($stdout)
+```
+
+The default reporter. Keep that in mind, we'll come back to this later.
+
+```Ruby
+load_path:         $LOAD_PATH,
+requires:          EMPTY_ARRAY,
+pathname:          Pathname,
+```
+
+What I like here is how configurable `Mutant`. You can even inject another $LOAD_PATH, additional requires or another Pathname'esque library.
+
+**The big picture for now:**
+
+TODO: proper picture
+
+bin -> cli -> config
```

Comment:
Ah, very helpful, thanks!
I actually intend to draw a real diagram, so not just ascii art.
Updated now and will use it later for the diagram.